### ISSUE TITLE
'my' variable scope incorrect in complex 'if' conditon

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6100]
Migrated from [rt.perl.org#18700](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=18700) (status was 'resolved')

Searchable as RT18700$

**Comment by p5pRT** [comment]
### From david_wallis@bmc.com


This is a bug report for perl from david\_wallis@&#8203;bmc\.com\,
generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.




---
The scope of a 'my' variable is supposed to go from the 'my' 
declaration to the end of the enclosing block\. In the following
example\, it apparently does not do this\. There appears to be a
gap in scope such that following clauses in a compound condition
do not receive the correct value\.

Failing example&#8203;:
$ /usr/bin/perl \-e 'if \(defined\(my $uid = getpwnam "NetCmmnd"\) && $uid ne 199\) \{
> warn "NetCmmnd already assigned to $uid\\n" \}'
NetCmmnd already assigned to 199

This example assumes the following entry in /etc/passwd&#8203;:
$ getent passwd NetCmmnd
NetCmmnd&#8203;:x&#8203;:199&#8203;:100&#8203;:PATROL EM&#8203;:/home/u/NetCmmnd&#8203;:/bin/csh

The same code without the 'my' declaration works&#8203;:
$ /usr/bin/perl \-e 'if \(defined\($uid = getpwnam "NetCmmnd"\) && $uid ne 199\) \{
> warn "NetCmmnd already assigned to $uid\\n" \}'
$
Also\, putting the my declaration in a separate statement before
the 'if' works&#8203;:
$ /usr/bin/perl \-e 'my $uid;
> if \(defined\($uid = getpwnam "NetCmmnd"\) && $uid ne 199\) \{
> warn "NetCmmnd already assigned to $uid\\n" \}'
$
The code should probably be using a numeric comparison on $uid\,
but that doesn't fix the problem&#8203;:
$/usr/bin/perl \-e 'if \(defined\(my $uid = getpwnam "NetCmmnd"\) && $uid \!= 199\) \{
>   warn "NetCmmnd already assigned to $uid\\n" \}'
NetCmmnd already assigned to 199
$
Obviously\, the $uid that is being tested after the && is not the same
$uid that is being assigned to the 'my' variable and printed in the 'warn'\.



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl v5\.6\.1&#8203;:

Summary of my perl5 \(revision 5\.0 version 6 subversion 1\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.9\, archname=sun4\-solaris\-64int
&nbsp;    uname='sunos localhost 5\.9 sun4u sparc sunw\,ultra\-1'
&nbsp;    config\_args=''
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=define use64bitall=undef uselongdouble=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-xO3 \-xdepend'\,
&nbsp;    cppflags=''
&nbsp;    ccversion='Sun WorkShop'\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=87654321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, usemymalloc=n\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =''
&nbsp;    libpth=/lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-ldl \-lm \-lc
&nbsp;    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lc
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-R /usr/perl5/5\.6\.1/lib/sun4\-solaris\-64int/CORE'
&nbsp;    cccdlflags='\-KPIC'\, lddlflags='\-G'

Locally applied patches&#8203;:
&nbsp;     9676 Port the OpenBSD glob\(\) security patch
&nbsp;     9678 Addendum to \#9676&#8203;: some missing changes from OpenBSD glob\.c
&nbsp;     9679 Up $File&#8203;::Glob&#8203;::VERSION\, add OpenBSD glob version note
&nbsp;     9693 $VERSION and Version\(\) on same line provokes CPAN\.pm warning
&nbsp;     9706 \#7210 broke \.packlist generation
&nbsp;     9707 ExtUtils&#8203;::Installed doesn't quote regex metacharacters in paths
&nbsp;     9775 Typo in utf8\.h
&nbsp;     9950 Revert integration of \#8254\,\#8255 in \#8620 \(causes coredump\)
&nbsp;    10021 Insecure regexes
&nbsp;    10091 $ref1 == $ref2 behaves unpredictably if not NV\_PRESERVES\_UV
&nbsp;    10093 Incorrect line numbers in AutoSplit
&nbsp;    10100 \[20010514\.027\] PL\_last\_in\_gv may not be GV if stale filehandle
&nbsp;    10145 \[20010515\.004\] Segfaults from premature GC
&nbsp;    10203 Don't think about UTF8
&nbsp;    10250 \[20010422\.005\] perl \-e '\{s//$\{\}/; //\}' segfaults
&nbsp;    10394 Leakage of file scope lexicals into predeclared subroutines
&nbsp;    10404 eval\.t was relying on pre\-\#10394 buggy behavior
&nbsp;    10412 Rationalize locale handling to fix bugs uncovered by \#10394
&nbsp;    10422 Potential buffer overrun if the radix separator > 1 byte
&nbsp;    10448 Lexicals outside eval weren't resolved correctly pre\-\#10394
&nbsp;    10450 Optimize \#10448 slightly
&nbsp;    10543 Add LC\_MESSAGES constant to POSIX module
&nbsp;    10667 \#10449 broke visibility of lexicals inside DB&#8203;::DB\(\)
&nbsp;    10739 C\<eval "/x$\\r\\n/x"> fails to compile correctly
&nbsp;    10939 Proposed fix for Pod&#8203;::Man
&nbsp;    11169 Doc patch for Tie&#8203;::Hash
&nbsp;    11374 Make h2ph grok ccsymbols fo the form 1234L\, 1234ULL etc
&nbsp;    11427 t/harness wasn't picking up all the tests
&nbsp;    11428 run/runenv\.t needs fflushNULL sanity
&nbsp;    11431 pod/\*\.t tests not picked up by t/TEST either
&nbsp;    11510 eval 'format foo=' would loop indefinitely
&nbsp;    11713 UTF8 wasn't printing for PVMGs
&nbsp;    11716 UTF8 flag should be meaningful only when POK
&nbsp;    11808 \[20010831\.002\] Bug in Term&#8203;::Cap on Solaris ansi terminal
&nbsp;    11847 Typo in perl\_clone\(\) code causes local\(\*foo\) breakage
&nbsp;    12005 \[20010912\.007\] substr reference core dump
&nbsp;    12024 Fix local\(\) precedence bug in \#8311
&nbsp;    12303 Fix 'local $\!=0;undef\*STDOUT;' segfault
&nbsp;    12304 Pod&#8203;::Html makes a poor guess at author
&nbsp;    12350 Typo in IO&#8203;::Seekable doc
&nbsp;    12496 Carp&#8203;::shortmess\_heavy\(\) doesn't notice trailing newline
&nbsp;    12549 readline\(\) doesn't work with 'our' variables
&nbsp;    12550 \#12549 wasn't aware of strictures
&nbsp;    12752 croak\(Nullch\) wasn't printing the contents of ERRSV
&nbsp;    12811 \[20011101\.069\] \\stat\('\.'\) gives 'free unref scalar' error
&nbsp;    12812 Slight modification of \#12811
&nbsp;    13149 Integrate \#13147 from mainline \(fixes nit in \#10091\)
&nbsp;    13261 Integrate \#8340\,\#13260 from mainline



---
@&#8203;INC for perl v5\.6\.1&#8203;:
&nbsp;    /usr/perl5/5\.6\.1/lib/sun4\-solaris\-64int
&nbsp;    /usr/perl5/5\.6\.1/lib
&nbsp;    /usr/perl5/site\_perl/5\.6\.1/sun4\-solaris\-64int
&nbsp;    /usr/perl5/site\_perl/5\.6\.1
&nbsp;    /usr/perl5/site\_perl
&nbsp;    /usr/perl5/vendor\_perl/5\.6\.1/sun4\-solaris\-64int
&nbsp;    /usr/perl5/vendor\_perl/5\.6\.1
&nbsp;    /usr/perl5/vendor\_perl
&nbsp;    \.



---
Environment for perl v5\.6\.1&#8203;:
&nbsp;    HOME=/
&nbsp;    LANG=en\_CA\.ISO8859\-1
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_ALL=
&nbsp;    LC\_COLLATE=en\_US\.ISO8859\-1
&nbsp;    LC\_CTYPE=en\_US\.ISO8859\-1
&nbsp;    LC\_MESSAGES=CommandPost
&nbsp;    LC\_MONETARY=en\_US\.ISO8859\-1
&nbsp;    LC\_NUMERIC=en\_US\.ISO8859\-1
&nbsp;    LC\_TIME=en\_US\.ISO8859\-1
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/sbin&#8203;:/usr/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/sbin/sh

**Comment by p5pRT** [comment]
### From rick.delaney@rogers.com

"david\_wallis@&#8203;bmc\.com \(via RT\)" \<perlbug@&#8203;perl\.org> writes&#8203;:

> Failing example&#8203;:
> $ /usr/bin/perl \-e 'if \(defined\(my $uid = getpwnam "NetCmmnd"\) && $uid ne 199\) \{
> > warn "NetCmmnd already assigned to $uid\\n" \}'
> NetCmmnd already assigned to 199

As documented in perlsub&#8203;:

&nbsp;       The declared variable is not introduced \(is not visible\)
&nbsp;       until after the current statement\.  Thus\,

&nbsp;           my $x = $x;

&nbsp;       can be used to initialize a new $x with the value of the
&nbsp;       old $x\, and the expression

&nbsp;           my $x = 123 and $x == 123

&nbsp;       is false unless the old $x happened to have the value
&nbsp;       "123"\.

> Obviously\, the $uid that is being tested after the && is not the same
> $uid that is being assigned to the 'my' variable and printed in the 'warn'\.

Exactly\.

\-\- 
Rick Delaney
rick\.delaney@&#8203;rogers\.com

**Comment by p5pRT** [comment]
### From david_wallis@bmc.com

Since perl does not DWIM in this case\, I consider it a bug that needs
correction\. It does DWIM in printing the correct value in the warn\, and it
does DWIM in the analgous situation of 'for my $x \.\.\.'\.

I can see that in 'my $x = $x;' and even 'my $x = 123 && $x == 123;' it is
reasonable to use the old value of $x\, but in 'my $x = 123 and $x == 123;'
this is unreasonable as the 'and' has low priority\, making the parse be '\(my
$x = 123\) and \($x == 123\);' which clearly puts the second $x "after" the
scope of the 'my' declaration\, and similarly for '\(my $x = 123\) && \($x ==
123\);' where the extent of the my 'statement' is explicitly set off by
parens\. 

'statement' is a somewhat nebulous concept in these cases of nesting\, but
DWIM has always been the perl way\, and this is clearly a surprise\.

David Wallis

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Rick Delaney \[mailto&#8203;:perlbug@&#8203;perl\.org\]
Sent&#8203;: Tuesday\, November 26\, 2002 7&#8203;:02 PM
To&#8203;: david\_wallis@&#8203;bmc\.com
Subject&#8203;: Re&#8203;: \[perl \#18700\] 'my' variable scope incorrect in complex 'if'
conditon


"david\_wallis@&#8203;bmc\.com \(via RT\)" \<perlbug@&#8203;perl\.org> writes&#8203;:

> Failing example&#8203;:
> $ /usr/bin/perl \-e 'if \(defined\(my $uid = getpwnam "NetCmmnd"\) && $uid ne
199\) \{
> > warn "NetCmmnd already assigned to $uid\\n" \}'
> NetCmmnd already assigned to 199

As documented in perlsub&#8203;:

&nbsp;       The declared variable is not introduced \(is not visible\)
&nbsp;       until after the current statement\.  Thus\,

&nbsp;           my $x = $x;

&nbsp;       can be used to initialize a new $x with the value of the
&nbsp;       old $x\, and the expression

&nbsp;           my $x = 123 and $x == 123

&nbsp;       is false unless the old $x happened to have the value
&nbsp;       "123"\.

> Obviously\, the $uid that is being tested after the && is not the same
> $uid that is being assigned to the 'my' variable and printed in the
'warn'\.

Exactly\.

\-\- 
Rick Delaney
rick\.delaney@&#8203;rogers\.com

**Comment by p5pRT** [comment]
### From islaw@adexec.com

Rick Delaney wrote&#8203;:
> "david\_wallis@&#8203;bmc\.com \(via RT\)" \<perlbug@&#8203;perl\.org> writes&#8203;:
>
>> Failing example&#8203;:
>> $ /usr/bin/perl \-e 'if \(defined\(my $uid = getpwnam "NetCmmnd"\) &&
>> $uid ne 199\) \{
>>> warn "NetCmmnd already assigned to $uid\\n" \}'
>> NetCmmnd already assigned to 199
>
> As documented in perlsub&#8203;:
>
>        The declared variable is not introduced \(is not visible\)
>        until after the current statement\.  Thus\,
>
>            my $x = $x;
>
>        can be used to initialize a new $x with the value of the
>        old $x\, and the expression
>
>            my $x = 123 and $x == 123
>
>        is false unless the old $x happened to have the value
>        "123"\.
>
>> Obviously\, the $uid that is being tested after the && is not the same
>> $uid that is being assigned to the 'my' variable and printed in the
>> 'warn'\.
>
> Exactly\.

Agreed\. Here are a couple quick test I ran \(using Perl 5\.6\.1\) that
concur \(as does my version of the perlsub doc\)&#8203;:
\[Copy pasted\, and cmd prompt text truncated\]

$ perl \-e 'print int \(my $x = 123 and $x == 123\)\. "\\n";'
0
$ perl \-e 'print int \(my $x = 123 and $x \!= 123\)\. "\\n";'
1
$ perl \-e 'print int \(my $x = 123 and $x == undef\)\. "\\n";'
1

\-\-
Islaw

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From david_wallis@bmc.com

Sorry I didn't answer immediately\. I've been on vacation\.

I take it that your agreement means that you think the current code and
documentation is reasonable\. I disagree for the following reasons&#8203;:

1\) The documentation quoted below says that the declared variable is not
visible 'until after the current statement'\. In the case of a 'my'
declaration nested inside an expression\, what should be considered the
extent of the 'current statement'?

One might reasonably expect that the current statement ends with the next
';'\, but this is obviously not the rule that perl follows in the case of a
'for' or 'if' statement\, because the new variable is visible inside the
block\(s\) that are nested within the 'current statement'\.

One might also reasonably expect that what is meant by 'current statement'
is the 'my' declaration\, assuming that it was coded as a standalone
statement \(with or without the optional assignment\)\. This is also not what
perl does\, because in the example&#8203;:
&nbsp;    my $x = 123 and $x == 123;
the extent of the declaration ends before the 'and' \(because of the low
precedence of 'and'\)\, yet 'and $x == 123' is treated as part of the
declaration statement and the $x that is compared to 123 is the old $x \(if
any\)\.

What is not reasonable is what perl does&#8203;: it ends the 'current statement' at
some arbitrary point depending upon the construct that the my declaration is
nested in\. For an 'if'\, the 'current statement' apparently ends before the
'then' block\, and the then and else blocks are in the scope of the
declaration\, but the condition is not in scope\, even if part of it follows
outside the extent of the 'my' declaration\. In a 'for \(my $x = \.\.\.; \.\.\.;
\.\.\.\.\.\.\) \{\.\.\.\}' statement\, the declared variable is in scope not only in the
'for' block\, but also the termination and increment expressions of the 'for'
statement\. I'm not sure what rules are followed for a 'my' declaration
within the termination or increment expressions\, but the documentation
clearly never envisions such a possibility\, and I doubt the current code
does either\.

2\) In making use of the term 'current statement' without acknowledging any
of the complications inherent therein\, the quoted documentation appears to
presume that a 'my' declaration is analogous to a C variable declaration
with an initializer\. But the 'my' declaration is a far more subtle beast
than that\. A C variable declaration is a statement that can only appear at
the head of a block\, but a 'my' declaration is not a statement\, it is a
builtin function that returns an lvalue\, and can appear anywhere that an
lvalue can\. The example&#8203;:
&nbsp;    my $x = 123 and $x == 123;
is parsed as&#8203;:
&nbsp;    \(\(my $x\) = 123\) and \($x == 123\);
and it is executed \(roughly\) as&#8203;:
&nbsp;    1\) Construct \(or lookup\) the rvalue of the assignment \(123\);
&nbsp;    2\) Construct \(or lookup\) the lvalue of the assignment \(my $x\);
&nbsp;    3\) Assign the rvalue \(123\) to the lvalue \($x\);
&nbsp;    4\) Test the return value of the assignment \(123\) as a boolean\,
&nbsp;        and skip the following expression if false;
&nbsp;      5\) Construct \(or lookup\) the left side of the == comparison \($x\);
&nbsp;      6\) Construct \(or lookup\) the right side of the == comparison
\(123\);
&nbsp;      7\) Make the comparison in a void context \(because this is a
nonsense example\)\.
&nbsp;    8\) Only now\(\!\) introduce the new variable \($x\) into scope\.
I assume assignment evaluates the rvalue before the lvalue because it is
right associative\. Similarly\, I assume == comparison evaluates left to right
because it is nonassociative\, but these are only guesses\, I don't code
depending upon these orders\.

The point I want to make is that 'until after the current statement' assumes
not only that the 'current statement' is a well\-define concept\, but also
that 'until after' is equally well defined\. Because a 'my' declaration has
an executable component\, 'until after' needs to be defined in terms of
evaluation order\. In the evaluation order given above\, it is obvious that
step 8 is well after the natural point where the new variable should be
introduced \(anywhere after step 1 and before step 4 would be natural\)\. Steps
4 through 7 have nothing to do with the 'my' declaration\, because they are
outside the extent of any of its component subexpressions\. Why should there
be such a 'hole' in the scope of $x?

In short\, I advocate two amendments to the perlsub doc\, and corresponding
changes to the perl code\.

1\) perlsub should acknowledge that 'my' is not a statement\, it is a builtin
function that produces an lvalue\. As such\, it can appear in the middle of
expressions\, and the new scope should begin as soon as the end of the rvalue
expression if it is the lvalue of an assignment\, and the end of its argument
expression \(also an lvalue\) otherwise\.

2\) the 'my' declaration affects the lexical scope of everything 'after' the
optional rvalue expression\, to the end of the current block\, where 'after'
is to be understood in terms of evaluation order\, not textual order\. In the
following example&#8203;:
&nbsp;    $x \.= $y    if my $y = \<>;
$x should receive the new value of $y appended to it\, not the old value
because the condition of the 'if' is evaluated before the concatencation\.

This should also have a beneficial effect of simplification on the perl
code\, because the proposed rule is much simpler than the current rule\.

David Wallis

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Islaw R\. \[mailto&#8203;:perlbug@&#8203;perl\.org\]
Sent&#8203;: Wednesday\, November 27\, 2002 9&#8203;:08 PM
To&#8203;: david\_wallis@&#8203;bmc\.com
Subject&#8203;: Re&#8203;: \[perl \#18700\] 'my' variable scope incorrect in complex 'if'
conditon


Rick Delaney wrote&#8203;:
> "david\_wallis@&#8203;bmc\.com \(via RT\)" \<perlbug@&#8203;perl\.org> writes&#8203;:
>
>> Failing example&#8203;:
>> $ /usr/bin/perl \-e 'if \(defined\(my $uid = getpwnam "NetCmmnd"\) &&
>> $uid ne 199\) \{
>>> warn "NetCmmnd already assigned to $uid\\n" \}'
>> NetCmmnd already assigned to 199
>
> As documented in perlsub&#8203;:
>
>        The declared variable is not introduced \(is not visible\)
>        until after the current statement\.  Thus\,
>
>            my $x = $x;
>
>        can be used to initialize a new $x with the value of the
>        old $x\, and the expression
>
>            my $x = 123 and $x == 123
>
>        is false unless the old $x happened to have the value
>        "123"\.
>
>> Obviously\, the $uid that is being tested after the && is not the same
>> $uid that is being assigned to the 'my' variable and printed in the
>> 'warn'\.
>
> Exactly\.

Agreed\. Here are a couple quick test I ran \(using Perl 5\.6\.1\) that
concur \(as does my version of the perlsub doc\)&#8203;:
\[Copy pasted\, and cmd prompt text truncated\]

$ perl \-e 'print int \(my $x = 123 and $x == 123\)\. "\\n";'
0
$ perl \-e 'print int \(my $x = 123 and $x \!= 123\)\. "\\n";'
1
$ perl \-e 'print int \(my $x = 123 and $x == undef\)\. "\\n";'
1

\-\-
Islaw

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
